tests/rule_test.php

```diff
@@ -0,0 +1,315 @@
+            $this->build_handler([
+                'availability' => ['available' => ai_manager_utils::AVAILABILITY_AVAILABLE],
+                'purposes' => [
+                    [
+                        'purpose' => 'feedback',
+                        'available' => ai_manager_utils::AVAILABILITY_DISABLED,
+                        'errormessage' => 'ai_manager->errormessage',
+                    ],
+                    ['purpose' => 'translate', 'available' => ai_manager_utils::AVAILABILITY_AVAILABLE],
+                ],
+            ])
+        );
+
+        $expected = 'ai_manager->errormessage';
+        $this->assertSame($expected, $rule->prevent_access());
+        $this->assertSame($expected, $rule->prevent_new_attempt(0, null));
+    }
+
+    /**
+     * Verifies multiple purpose messages are combined.
+     *
+     * @covers ::prevent_access
+     * @covers ::prevent_new_attempt
+     */
+    public function test_prevent_access_combines_multiple_purpose_messages(): void {
+        $rule = $this->create_rule(
+            $this->build_handler([
+                'availability' => ['available' => ai_manager_utils::AVAILABILITY_AVAILABLE],
+                'purposes' => [
+                    [
+                        'purpose' => 'feedback',
+                        'available' => ai_manager_utils::AVAILABILITY_HIDDEN,
+                        'errormessage' => '',
+                    ],
+                    [
+                        'purpose' => 'translate',
+                        'available' => ai_manager_utils::AVAILABILITY_DISABLED,
+                        'errormessage' => 'translate disabled',
+                    ],
+                ],
+            ])
+        );
+
+        $expected = \html_writer::alist([
+            get_string('error_aipurposeunavailable', 'quizaccess_ai', 'feedback'),
+            'translate disabled',
+        ]);
+
+        $this->assertSame($expected, $rule->prevent_access());
+        $this->assertSame($expected, $rule->prevent_new_attempt(0, null));
+    }
+
+    /**
+     * Creates the rule and injects the provided handler.
+     *
+     * @param ai_access_handler $handler
+     * @return \quizaccess_ai
+     */
+    protected function create_rule(ai_access_handler $handler): \quizaccess_ai {
+        \core\di::set(ai_access_handler::class, $handler);
+        $this->setAdminUser();
+
+        $quizobj = $this->createMock(quiz_settings::class);
+        $quizobj->method('get_quiz')->willReturn((object)[
+            'id' => 1,
+            'timeclose' => 0,
+            'timelimit' => 0,
+        ]);
+        $quizobj->method('get_context')->willReturn(\context_system::instance());
+
+        return new \quizaccess_ai($quizobj, time());
+    }
+
+    /**
+     * Builds a handler stub that always returns the provided AI configuration.
+     *
+     * @param array $config
+     * @return ai_access_handler
+     */
+    protected function build_handler(array $config): ai_access_handler {
```

Comment:
was the whole idea of creating the `ai_access_handler` class not that you can just mock it?